File: lib/ClientUdp/command/send_handler.dart
import 'dart:io';

import 'command_handler.dart';
import '../client_state.dart';
import '../constants.dart';
import '../handshake_manager.dart';
import '../json_helper.dart';

class SendHandler implements CommandHandler {
  @override
  void handle(String args, ClientState clientState, HandshakeManager handshakeManager) {
    List<String> parts = args.split(RegExp(r'\s+'));
    List<String> sendArgs = [];
    
    if (parts.isNotEmpty) {
      sendArgs.add(parts[0]);
      if (parts.length > 1) {
        sendArgs.add(parts.sublist(1).join(' '));
      }
    }
    
    if (sendArgs.length != 2) {
      print("Usage: ${Constants.CMD_SEND} <room_id> <message>");
      stdout.write("> ");
      return;
    }

    String roomId = sendArgs[0];
    String content = sendArgs[1];

    if (clientState.sessionKey == null) {
      print("You must be logged in to send messages. Use /login <id> <pw>");
      stdout.write("> ");
      return;
    }

    // Basic validation, server should do more thorough checks
    if (roomId.trim().isEmpty || content.isEmpty) {
      print("Usage: ${Constants.CMD_SEND} <room_id> <message>");
      stdout.write("> ");
      return;
    }

    Map<String, dynamic> data = {
      Constants.KEY_CHAT_ID: clientState.currentChatId,
      Constants.KEY_ROOM_ID: roomId.trim(),
      Constants.KEY_CONTENT: content
    };
    
    Map<String, dynamic> request = JsonHelper.createRequest(Constants.ACTION_SEND_MESSAGE, data);
    handshakeManager.sendClientRequestWithAck(request, Constants.ACTION_SEND_MESSAGE, clientState.sessionKey!);
    // No need to print "> " here
  }

  @override
  String getDescription() {
    return Constants.CMD_SEND_DESC;
  }
}

File: UdpClientFactor/src/main/java/UdpChatClient/command/SendHandler.java
package UdpChatClient.command;

import com.google.gson.JsonObject;

import UdpChatClient.ClientState;
import UdpChatClient.Constants;
import UdpChatClient.HandshakeManager;
import UdpChatClient.JsonHelper;

public class SendHandler implements CommandHandler {

    @Override
    public void handle(String args, ClientState clientState, HandshakeManager handshakeManager) {
        String[] sendArgs = args.split("\\s+", 2);
        if (sendArgs.length != 2) {
            System.out.println("Usage: " + Constants.CMD_SEND + " <room_id> <message>");
            System.out.print("> ");
            return;
        }

        String roomId = sendArgs[0];
        String content = sendArgs[1];

        if (clientState.getSessionKey() == null) {
            System.out.println("You must be logged in to send messages. Use /login <id> <pw>");
            System.out.print("> ");
            return;
        }

        // Basic validation, server should do more thorough checks
        if (roomId.trim().isEmpty() || content.isEmpty()) {
            System.out.println("Usage: " + Constants.CMD_SEND + " <room_id> <message>");
            System.out.print("> ");
            return;
        }

        JsonObject data = new JsonObject();
        data.addProperty(Constants.KEY_CHAT_ID, clientState.getCurrentChatId());
        data.addProperty(Constants.KEY_ROOM_ID, roomId.trim());
        data.addProperty(Constants.KEY_CONTENT, content);
        JsonObject request = JsonHelper.createRequest(Constants.ACTION_SEND_MESSAGE, data);
        handshakeManager.sendClientRequestWithAck(request, Constants.ACTION_SEND_MESSAGE, clientState.getSessionKey());
        // No need to print "> " here
    }

    @Override
    public String getDescription() {
        return Constants.CMD_SEND_DESC;
    }
}
